packages/lib/src/utils/useForm.ts

```diff
@@ -0,0 +1,107 @@
+import { useEffect, useMemo, useState } from 'preact/hooks';
+import Validator, { ValidatorRules } from './Validator/FormValidator';
+
+function useForm<DataState = { [key: string]: any }>(props: { rules?: ValidatorRules; [key: string]: any }) {
+    const { rules = {}, formatters = {}, defaultData = {} } = props;
+    const validator = new Validator(rules);
+    const [schema, setSchema] = useState<string[]>(props.schema ?? []);
+    const [errors, setErrors] = useState<any>({});
+    const [valid, setValid] = useState<any>({});
+    const [data, setData] = useState<DataState>({} as DataState);
+    const isValid = useMemo(() => schema.reduce((acc, val) => acc && valid[val], true), [valid]);
+
+    const updateFieldData = (key, value) => {
+        setData(prevData => ({ ...prevData, [key]: value }));
+    };
+
+    const updateFieldValidation = (key, validation) => {
+        setValid(prevValid => ({ ...prevValid, [key]: validation.isValid ?? false }));
+        setErrors(prevErrors => ({ ...prevErrors, [key]: validation.hasError() ? validation.getError() : false }));
+    };
+
+    /**
+     * Format and validate a field
+     */
+    const processField = (key, value, mode) => {
+        const formattedValue = formatters[key] ? formatters[key](value ?? '') : value;
+        const validationResult = validator.validate(key, formattedValue, mode);
+        return [formattedValue, validationResult];
+    };
+
+    const reindexSchema = keys => {
+        const cleanupRemovedFields = (prevData, initialValue) =>
+            keys.reduce((acc, key) => {
+                acc[key] = prevData[key] !== undefined ? prevData[key] : initialValue;
```

Comment:
Would this work on this case? Or could it be that `prevData[key]` is `null` and we want to keep it that way?
```suggestion
                acc[key] = prevData[key] ?? initialValue;
```

Reply:
That's the thing, not sure how safe it would be to make that kind of assumption. That's why it's currently just setting it to the `initialValue` if it's not defined at all.